lib/functions/compilation/kernel-debs.sh

```diff
@@ -444,10 +444,13 @@ function kernel_package_callback_linux_headers() {
 	# Understand: I'm sending the logs of this to the bitbucket ON PURPOSE: "clean" tries to use clang, ALSA, etc, which are not available.
 	#             The logs produced during this step throw off developers casually looking at the logs.
 	#             Important: if the steps _fail_ here, you'll have to enable DEBUG=yes to see what's going on.
+	#
+	# In order for the cleanup to be correct  for tools, we need to pass the VMLINUX_BTF variable,
+	# which contains the real path to the newly compiled vmlinux file.
 	declare make_bitbucket="&> /dev/null"
 	[[ "${DEBUG}" == "yes" ]] && make_bitbucket=""
 	run_host_command_logged cd "${headers_target_dir}" "&&" make "ARCH=${SRC_ARCH}" "M=scripts" clean "${make_bitbucket}"
-	run_host_command_logged cd "${headers_target_dir}/tools" "&&" make "ARCH=${SRC_ARCH}" clean "${make_bitbucket}"
+	run_host_command_logged cd "${headers_target_dir}/tools" "&&" make "ARCH=${SRC_ARCH}" "VMLINUX_BTF=${kernel_work_dir}/vmlinux" clean "${make_bitbucket}"
```

Comment:
Why? 
Seems to me you hit an edge case where for some reason the "clean" target (of standard Kbuild) is going into bpf stuff which tries to find btf which was not built. Check your `.config` for sanity?

Does this happen to any family other than sunxi?

Reply:
Alternatively you found a bug in Kbuild, so please report it upstream.

Reply:
> Does this happen to any family other than sunxi?

[patch/kernel/archive/rockchip64-6.12/0001-tools-disable-sched_ext_clean.patch](https://github.com/armbian/build/blob/main/patch/kernel/archive/rockchip64-6.12/0001-tools-disable-sched_ext_clean.patch)
For rockchip, the problem is solved using this crutch.
I can add it for sunxi and we will close the problem.

There are no problems in the standard Kbuild.
At least I don't see them.
But we use a non-standard and our own Armbian algorithm for assembling the kernel header package.
Since version 6.12, the standard algorithm in the kernel has changed for headers. Now it differs for native and cross compilation.
We need to take this into account somehow.

Reply:
> Check your .config for sanity?

What did you see in this?

Reply:
> [patch/kernel/archive/rockchip64-6.12/0001-tools-disable-sched_ext_clean.patch](https://github.com/armbian/build/blob/main/patch/kernel/archive/rockchip64-6.12/0001-tools-disable-sched_ext_clean.patch?rgh-link-date=2024-12-28T08%3A47%3A28Z)
> For rockchip, the problem is solved using this crutch.

Interesting. This has flown under my radar. It is in rk64 6.12 and in `sm8250-6.12`. 
We don't have this for meson64, though, and it builds / cleans fine, on 6.12.y -- how can that be? (I imagine it can only be something in the .config, that's why I asked: Maybe some CONFIG_xx ends up enabling the whole BTF).

Either way, seems like we're not copying enough of `tools/` into the working copy so that `clean` can work for sched_ext. I could look into that and solve it across the board -- I will dedicate some time to this ASAP.


> Since version 6.12, the standard algorithm in the kernel has changed for headers

Indeed, but we're not using it. After more than 10 years, Mr. Masashiro added the capability that was missing and, mostly, the reason Armbian was created and existed for so long: https://github.com/torvalds/linux/commit/f1d87664b82aeeaa1be9ee22dc85a59fd5a60d63 -- hopefully in some 2-3 years 6.12.y will be the lowest common denominator and we can get rid of all the cross-compiling-headers code pyrotechnics.

Reply:
I am currently finishing handling the build of armhf on arm64-but-not-armhf hosts (Apple Silicon vs binfmt vs qemu-static) which is the first blocker for 32-bit boards; then I will fix the `kernel-dtb` bug, then I will look into this `tools/` vs BTF problem, in that order.

Reply:
Well, I cherry picked this PR. It builds fine.

I couldn't reproduce the problem in any way:

1) I dropped the `tools-disable-sched_ext_clean.patch` from rockchip64 kernel and it still built and packaged no problem (`rockpro64`)
2) I reverted your commit adding `VMLINUX_BTF` to the make clean invocation, and it still built no problem (`bananapim4zero`)

I tried on both x86 and arm64 build hosts.

Thus none of this makes any sense to me.

Reply:
Just delete the make clean under tools: https://github.com/armbian/build/pull/7625.

Reply:
See #7625 indeed for the discussion; simply not-calling make clean in tools results in binaries being shipped in linux-headers, and if cross-compiling, for the wrong architecture.

After considering all the factors at play here, I think @The-going 's proposed solution (passing a fake `VMLINUX_BTF=` var to make clean in tools) is the best one. I'll resolve this and approve the PR.

Thank you all so much for the learnings here and sorry for the back and forth's.